go/mysql/conn.go

```diff
@@ -1206,25 +1236,127 @@ func (c *Conn) handleNextCommand(handler Handler) error {
 			}
 		}
 
+		c.discardCursor()
+
 		if err := c.writeOKPacket(0, 0, c.StatusFlags, 0); err != nil {
 			log.Error("Error writing ComStmtReset OK packet to client %v: %v", c.ConnectionID, err)
 			return err
 		}
-
 	case ComStmtFetch:
-		_, _, _ = c.parseComStmtFetch(data)
+		c.startWriterBuffering()
+		endFetch := func() error {
+			if err = c.writeEndResult(false, 0, 0, handler.WarningCount(c)); err != nil {
+				log.Errorf("Error writing result to %s: %v", c, err)
+				return err
+			}
+			if err = c.flush(); err != nil {
+				log.Errorf("Conn %v: Flush() failed: %v", c.ID(), err)
+				return err
+			}
+			return nil
+		}
+
+		stmtID, numRows, ok := c.parseComStmtFetch(data)
 		c.recycleReadPacket()
+		if !ok {
+			log.Error("Got unhandled packet from client %v, returning error: %v", c.ConnectionID, data)
+			if werr := c.writeErrorPacket(ERUnknownComError, SSUnknownComError, "error handling packet: %v", data); werr != nil {
+				log.Error("Error writing error packet to client: %v", werr)
+				return werr
+			}
+		}
 
-		err = fmt.Errorf("COM_STMT_FETCH is not suppported")
-		if werr := c.writeErrorPacketFromError(err); werr != nil {
-			// If we can't even write the error, we're done.
-			log.Errorf("Error writing query error to %s: %v", c, werr)
-			return werr
+		// cursor is closed
+		if c.cs == nil {
+			c.StatusFlags &= ^uint16(ServerCursorExists)
+			c.StatusFlags |= uint16(ServerCursorLastRowSent)
+			if err = endFetch(); err != nil {
+				return err
+			}
+			return nil
 		}
 
+		// cursor is open
+		c.StatusFlags |= uint16(ServerCursorExists)
+		c.StatusFlags &= ^uint16(ServerCursorLastRowSent)
+
+		// fetching from wrong statement
+		if stmtID != c.cs.stmtID {
+			log.Errorf("Requested stmtID does not match stmtID of open cursor. Client %v, returning error: %v", c.ConnectionID, data)
+			if werr := c.writeErrorPacket(ERUnknownComError, SSUnknownComError, "error handling packet: %v", data); werr != nil {
+				log.Error("Error writing error packet to client: %v", err)
+				return werr
+			}
+		}
+
+		// while we are still expecting rows
+		for c.cs.pending == nil || len(c.cs.pending.Rows) < int(numRows) {
+			// receive rows from next to cs.pending
+			var newqr *sqltypes.Result
+			newqr, ok = <-c.cs.next
+
+			// channel is closed, meaning there are no more rows
+			if !ok {
+				// check if there was an error
+				if err, ok = <- c.cs.done; ok {
+					close(c.cs.done)
```

Comment:
If we ever want this to close, we should have the sender close it. This server code should be structured to only ever successfully receive from it once though.